File: src/backend/src/controller/organization.rs
use crate::{
    dto::{
        CreateOrganizationRequest, CreateOrganizationResponse, DeleteOrganizationRequest,
        GetOrganizationRequest, GetOrganizationResponse, ListMyOrganizationsResponse,
        UpdateOrganizationRequest, UpdateOrganizationResponse,
    },
    service::organization_service,
};
use canister_utils::{assert_authenticated, ApiResultDto};
use ic_cdk::{api::msg_caller, *};

#[query]
fn list_my_organizations() -> ApiResultDto<ListMyOrganizationsResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    organization_service::list_my_organizations(&caller).into()
}

#[update]
fn create_organization(req: CreateOrganizationRequest) -> ApiResultDto<CreateOrganizationResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    organization_service::create_organization(&caller, req).into()
}

#[query]
fn get_organization(req: GetOrganizationRequest) -> ApiResultDto<GetOrganizationResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    organization_service::get_organization(&caller, req).into()
}

#[update]
fn update_organization(req: UpdateOrganizationRequest) -> ApiResultDto<UpdateOrganizationResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    organization_service::update_organization(&caller, req).into()
}

#[update]
fn delete_organization(req: DeleteOrganizationRequest) -> ApiResultDto<()> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    organization_service::delete_organization(&caller, req).into()
}

File: src/frontend/src/lib/api/organization.ts
import {
  mapListMyOrganizationsResponse,
  mapCreateOrganizationRequest,
  mapCreateOrganizationResponse,
  mapGetOrganizationRequest,
  mapGetOrganizationResponse,
  mapUpdateOrganizationRequest,
  mapUpdateOrganizationResponse,
  mapDeleteOrganizationRequest,
  mapDeleteOrganizationResponse,
  mapListOrgUsersRequest,
  mapListOrgUsersResponse,
  mapListOrganizationsRequest,
  mapListOrganizationsResponse,
  mapGetOrgBillingPlanRequest,
  mapGetOrgBillingPlanResponse,
  mapListMyOrgBillingPlansResponse,
  mapSetOrgBillingPlanRequest,
  mapSetOrgBillingPlanResponse,
  type ListMyOrganizationsResponse,
  type ListOrganizationsRequest,
  type ListOrganizationsResponse,
  type CreateOrganizationRequest,
  type OrganizationResponse,
  type GetOrganizationRequest,
  type UpdateOrganizationRequest,
  type DeleteOrganizationRequest,
  type ListOrgUsersRequest,
  type OrgUser,
  type GetOrgBillingPlanRequest,
  type MyOrgBillingPlan,
  type OrgBillingPlan,
  type SetOrgBillingPlanRequest,
} from '@/lib/api-models';
import type { ActorSubclass } from '@icp-sdk/core/agent';
import type { _SERVICE } from '@ssn/backend-api';

export class OrganizationApi {
  constructor(private readonly actor: ActorSubclass<_SERVICE>) {}

  public async listMyOrganizations(): Promise<ListMyOrganizationsResponse> {
    const res = await this.actor.list_my_organizations();
    return mapListMyOrganizationsResponse(res);
  }

  public async listOrganizations(
    req: ListOrganizationsRequest,
  ): Promise<ListOrganizationsResponse> {
    const res = await this.actor.admin_list_organizations(
      mapListOrganizationsRequest(req),
    );
    return mapListOrganizationsResponse(res);
  }

  public async createOrganization(
    req: CreateOrganizationRequest,
  ): Promise<OrganizationResponse> {
    const res = await this.actor.create_organization(
      mapCreateOrganizationRequest(req),
    );
    return mapCreateOrganizationResponse(res);
  }

  public async getOrganization(
    req: GetOrganizationRequest,
  ): Promise<OrganizationResponse> {
    const res = await this.actor.get_organization(
      mapGetOrganizationRequest(req),
    );
    return mapGetOrganizationResponse(res);
  }

  public async updateOrganization(
    req: UpdateOrganizationRequest,
  ): Promise<OrganizationResponse> {
    const res = await this.actor.update_organization(
      mapUpdateOrganizationRequest(req),
    );
    return mapUpdateOrganizationResponse(res);
  }

  public async deleteOrganization(
    req: DeleteOrganizationRequest,
  ): Promise<void> {
    const res = await this.actor.delete_organization(
      mapDeleteOrganizationRequest(req),
    );
    mapDeleteOrganizationResponse(res);
  }

  public async listOrgUsers(req: ListOrgUsersRequest): Promise<OrgUser[]> {
    const res = await this.actor.list_org_users(mapListOrgUsersRequest(req));
    return mapListOrgUsersResponse(res);
  }

  public async getOrgBillingPlan(
    req: GetOrgBillingPlanRequest,
  ): Promise<OrgBillingPlan> {
    const res = await this.actor.get_org_billing_plan(
      mapGetOrgBillingPlanRequest(req),
    );
    return mapGetOrgBillingPlanResponse(res);
  }

  public async setOrgBillingPlan(req: SetOrgBillingPlanRequest): Promise<void> {
    const res = await this.actor.admin_set_org_billing_plan(
      mapSetOrgBillingPlanRequest(req),
    );
    mapSetOrgBillingPlanResponse(res);
  }

  public async listMyOrgBillingPlans(): Promise<MyOrgBillingPlan[]> {
    const res = await this.actor.list_my_org_billing_plans();
    return mapListMyOrgBillingPlansResponse(res);
  }
}
